File: W-23/Assignments/asgn3/code/shell.c
#include "shell.h"

#include "gaps.h"

#include <assert.h>
#include <math.h>
#include <stdio.h>

void shell_sort(Stats *stats, uint32_t *A, uint32_t n) {
  for (uint32_t const *gptr = gaps; gptr != gaps + GAPS; ++gptr) {
    uint32_t gap = *gptr;
    for (uint32_t i = gap; i < n; i += 1) {
      uint32_t j = i;
      uint32_t temp = move(stats, A[i]);
      while (j >= gap && cmp(stats, temp, A[j - gap]) < 0) {
        A[j] = move(stats, A[j - gap]);
        j -= gap;
      }
      A[j] = move(stats, temp);
    }
  }
}

File: W-22/Assignments/asgn3/python/shell.py
from math import log

def gaps(n):
    for i in range(int(log(3 + 2 * n) / log(3)), 0, -1):
        yield (3**i - 1) // 2

def shell_sort(A: list):
    for gap in gaps(len(A)):
        for i in range(gap, len(A)):
            j = i
            temp = A[i]
            while j >= gap and temp < A[j - gap]:
                A[j] = A[j - gap]
                j -= gap
            A[j] = temp
